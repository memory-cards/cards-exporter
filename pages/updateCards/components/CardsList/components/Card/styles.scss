.card {
  background-color: #ffffff;
  margin: 0 5px 5px;
  padding: 5px;
  min-height: 40px;
  border-radius: 6px;
  box-shadow: 3px 3px 3px darkslategray;
  border-left: 3px solid transparent;

  &:last-child {
    margin-bottom: 3px;
  }

  &:hover {
    cursor: pointer;
    background-color: #eeeded;
    transition: all 0.5s;
  }

  .selected {
    border-left-color: darkred;
  }

  .tag {
    display: inline-block;
    background-color: lightgray;
    margin: 4px;
    border-radius: 6px;
    padding: 0px 5px;
  }
}
