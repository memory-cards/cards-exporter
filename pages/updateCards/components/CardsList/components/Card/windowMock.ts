window.document.createElement = jest.fn().mockImplementation(() => {
  const element = {};

  Object.defineProperty(element, "innerHTML", {
    /* tslint:disable object-literal-shorthand */
    set: function(x) {
      this.innerText = x;
    }
  });

  return element;
});
