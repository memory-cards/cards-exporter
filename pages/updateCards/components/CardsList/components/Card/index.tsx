import React from "react";
import { ICardDefinition } from "~/typings/ICardDefinition";

import "./styles.scss";

interface Props {
  card: ICardDefinition;
  selectCard: (selectedCard: ICardDefinition) => void;
}

const Card = ({ card, selectCard }: Props) => {
  const onClick = () => {
    selectCard(card);
  };

  const getShortCardQuestion = () => {
    const element = document.createElement("div");
    element.innerHTML = card.card.question;
    const text = element.innerText
      .split(" ")
      .slice(0, 10)
      .map(word => (word.length < 30 ? word : word.substr(0, 30)))
      .join(" ");
    return text.length < element.innerText.length ? `${text}...` : text;
  };

  return (
    <li className="card" onClick={onClick}>
      <div>{getShortCardQuestion()}</div>
      <br />
      <div>
        {card.tags.map(tag => (
          <span className="tag">{tag}</span>
        ))}
      </div>
    </li>
  );
};

export default Card;
