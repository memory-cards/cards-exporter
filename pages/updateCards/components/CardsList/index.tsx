import React from "react";
import { ICardDefinition } from "~/typings/ICardDefinition";

import Card from "./components/Card";

interface Props {
  cards: ICardDefinition[];
}

const CardsList = ({ cards }: Props) => {
  return (
    <ul>
      {cards.map(card => (
        <Card card={card} />
      ))}
    </ul>
  );
};

export default CardsList;
