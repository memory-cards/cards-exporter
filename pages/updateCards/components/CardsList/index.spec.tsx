/**
 * @jest-environment jsdom
 */
import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import * as TestRenderer from "react-test-renderer";

import "./components/Card/windowMock";

import Card from "./components/Card";
import CardsList from "./index";

const cards = [
  {
    card: {
      comment: "comment1",
      question: "question1"
    },
    lang: "lang1",
    tags: ["tag1"],
    type: "type"
  },
  {
    card: {
      comment: "comment2",
      question: "question2"
    },
    lang: "lang2",
    tags: ["tag2"],
    type: "type"
  }
];

describe("With Enzyme", () => {
  let cardsList: ShallowWrapper;

  beforeEach(() => {
    cardsList = shallow(
      <CardsList cards={cards} selectCard={jest.fn()} selectedCard={cards[0]} />
    );
  });

  describe("CardsList", () => {
    it("renders correct cards amount", () => {
      expect(cardsList.find(Card).length).toBe(cards.length);
    });

    it("marks selected card correctly", () => {
      const renderedCards = cardsList.find(Card);
      expect(renderedCards.at(0).prop("isSelected")).toBe(true);
      expect(renderedCards.at(1).prop("isSelected")).toBe(false);
    });
  });
});

describe("With Snapshot Testing", () => {
  it('IndexPage shows "Cards Exporter"', () => {
    const component = TestRenderer.create(
      <CardsList cards={cards} selectCard={jest.fn()} selectedCard={cards[0]} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
