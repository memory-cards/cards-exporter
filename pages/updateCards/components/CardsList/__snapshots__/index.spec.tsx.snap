// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Snapshot Testing IndexPage shows "Cards Exporter" 1`] = `
<ul>
  <li
    className="card selected"
    onClick={[Function]}
  >
    <div>
      question1
    </div>
    <br />
    <div>
      <span
        className="tag"
      >
        tag1
      </span>
    </div>
  </li>
  <li
    className="card"
    onClick={[Function]}
  >
    <div>
      question2
    </div>
    <br />
    <div>
      <span
        className="tag"
      >
        tag2
      </span>
    </div>
  </li>
</ul>
`;
